File: solver.py
#!/usr/bin/env python3
from guesses import guesses
from answers import answers
from time import time
both = guesses + answers

class Game:
    def __init__(self, first_guess, bot=False, n=1):
        # initialising with bot=True, makes the computer solve n words
        # initialising with bot=False, is user mode, where the user inputs the outcomes
        self.guess = first_guess
        self.bot = bot
        self.answers = answers
        self.n = 1
        self.outcomes = total_outcomes()
        if self.bot:
            self.time = time()
            self.chosen = []
            self.errors = 0
            self.attempts = 0
            self.words = 0
            self.total = n
            self.rand_word()

    def rand_word(self):
        # finding a random word in answers, making sure it has not been picked before
        from random import randint
        i = randint(0, len(self.answers) - 1)
        while i in self.chosen:
            i = randint(0, len(self.answers) - 1)
        self.chosen.append(i)
        self.word = self.answers[i]
        print(f'\n{self.word.upper()} - {self.words+1}')
    
    
    def check(self, guess):
            # checking a guess against the word, returns the outcome of the guess

            # default outcome
            outcome = ['-', '-', '-', '-', '-']

            # keeping track of the letters that have been checked (for doubles)
            letters = {}

            # get g's and add to letters
            for x in range(len(guess)):
                if guess[x] == self.word[x]:
                    outcome[x] = 'g'
                    try: 
                        letters[guess[x]] += 1
                    except:
                        letters[guess[x]] = 1
            
            # getting o's if there are any while taking into consideration doubles
            for x in range(len(guess)):
                count = self.word.count(guess[x])
                if count and outcome[x] != 'g':
                    try:
                        if letters[guess[x]] < count:
                            outcome[x] = 'o'
                            letters[guess[x]] += 1
                    except:
                        outcome[x] = 'o'
                        letters[guess[x]] = 1
            return outcome

    def solve_word(self):
        # function used to solve for a word

        # guess is printed (extra space is to make sure it clears the line when printed)
        print(f'GUESS : {self.guess.upper()}              ')
        self.n += 1

        # result of guess is obtained
        if self.bot:
            self.res = self.check(self.guess)
            print('RESULT : ' + ''.join(self.res))
        else:
            self.res = list(input(f"RESULT : "))
        
        # answers are refined to meet result
        self.new_ans = list(filter(lambda an: is_possible(an, self.guess, self.res), self.answers))
        print(f'ANSWERS LEFT : {len(self.new_ans)}')
        
        # if one answer is left, game is won, if none are left, there was an error
        if len(self.new_ans) in (0, 1):
            self.word_complete(bool(len(self.new_ans)))
            return
        
        # hold a temporary best
        best = ('', len(self.new_ans))

        # check each guess
        for g in range(len(both)):
            total = len(self.outcomes)
            possible_ans = 0
            for o in self.outcomes: # check each outcome

                # refine the answers to get the amount of words left
                ans_left = len(list(filter(lambda a: is_possible(a, both[g], o), self.new_ans)))
                if not ans_left:
                    total -= 1
                possible_ans += ans_left
            
            # the temporary best is compared to the current guess
            best = min(best, (both[g], possible_ans/total), key=lambda x: x[1])

            print(f'Checking {g+1}/{len(both)}', end='\r')

        # next guess is determined and solving continues
        self.answers = self.new_ans
        self.guess = best[0]
        self.solve_word()

    def word_complete(self, success):
        # to be run when a word is either solved or an error occured

        # print output to user and update variables
        if success:
            print(f'** WORD = {self.new_ans[0].upper()} - {self.n} Guesses **')
            if self.bot:
                self.attempts += self.n
            else:
                return
        elif not success and self.bot:
            print(f'ERROR OCCURED')
            self.errors += 1
        else:
            # if error occurs, print all possible answers before error occured (user mode)
            print(f'Error Occured, answer is one of : {self.answers}')
            return

        # end word guessing if total is reached, print summary (bot only)
        self.words += 1
        if self.words == self.total:
            print(f'\nCompleted {self.total} Words\nAverage Guesses: {round(self.attempts/(self.total-self.errors), 2)}\nErrors: {self.errors}\nTime : {round((time()-self.time)/60, 2)} mins')
            return
        
        # reinitialise variables, get a new word and solve it (bot only)
        self.n = 1
        self.answers = answers
        self.guess = 'trace'
        self.rand_word()
        self.solve_word()


def total_outcomes(n=5, part=[]):
    # function to get all possible outcomes
    if len(part) == n:
        return [part]
    else:
        res = []
        for o in ['-', 'o', 'g']:
            res += total_outcomes(n, part + [o])
        return res


def is_possible(word, guess, outcome):
    # determines whether a word is viable to be an answer, given an attempt and the outcome the attempt provided

    # split the indexes into groups [[g's], [o's], [-'s]] as checking requires g's to be checked first, o's second and -'s last
    order = [[],[],[]]
    for i in range(len(outcome)):
        if outcome[i] == 'g':
            order[0].append(i)
        elif outcome[i] == 'o':
            order[1].append(i)
        else:
            order[2].append(i)

    # keeping tracking of letters that have been checked (for doubles)
    checked = {}
    for group in order:
        for x in group: 
            # go through the g's in outcome first, then the o's, then the -'s

            count = word.count(guess[x])

            # if the outcome is g, check if word contains guess[x] at position x
            # add to checked
            if outcome[x] == 'g':
                if word[x] != guess[x]:
                    return False
                try:
                    checked[guess[x]] += 1
                except:
                    checked[guess[x]] = 1

            # if outcome is o, if checked[guess[x]] exists, it must be less than count and guess[x] cannot equal word[x] in order for the word to be viable
            # if checked[guess[x]] doesnt exist, there must be a count and guess[x] cannot equal word[x] for the word to be viable
            elif outcome[x] == 'o':
                try:
                    if checked[guess[x]] == count or guess[x] == word[x]:
                        return False
                    checked[guess[x]] += 1
                except:
                    if not count or guess[x] == word[x]:
                        return False
                    checked[guess[x]] = 1

            # if the outcome is -, if checked[guess[x]] exists and there is a count, checked[guess[x]] must already be equal to count for the word to be viable
            # if checked[guess[x]] doesnt exist but there is a count the word cannot be viable
            elif outcome[x] == '-':
                if count:
                    try:
                        if checked[guess[x]] < count:
                            return False
                    except:
                        return False
    return True


def get_first_guess():
    # code used to produce first_guess.txt and find that 'trace' was the best first word
    outcomes = total_outcomes()
    word_set = []
    for g in range(len(both)):
        total = len(outcomes)
        possible_ans = 0
        for o in outcomes: # check all outcomes

            # see how many answers are left are both[g] is guessed
            ans_left = len(list(filter(lambda a: is_possible(a, both[g], o), answers)))
            
            # if 0 answers are left, outcome is impossible therefore excluded from total
            if not ans_left:
                total -= 1
            possible_ans += ans_left

        # get average, print for reference and append to word_set
        av = possible_ans/total
        print(f'{g+1} - {both[g]} - {av}') 
        word_set.append((both[g], av))

    # create file to write the sorted word_set in
    with open('first_guess.txt', 'w') as f:
        for x in sorted(word_set, key=lambda w: w[1]):
            f.write(f'{x[0]} - {x[1]}\n')


if __name__ == '__main__':
    wordle = Game('trace', True)
    wordle.solve_word()

# current runtime = 4sec per answer left

File: solver.c
/**
 * Solver for Wordle
 * @author Samir Gupta
 * @date 2022-09-11
 * @note This implementation was originally translated from Python
 */

#include "shared.c"

#define ANSWER_LENGTH 2315   // the amount of valid answers wordle has
#define GUESSES_LENGTH 12972 // the amount of valid guesses wordle allows

#define WORD_SIZE 6 // the total word size of each word (5 + escaping)

#define TOTAL_OUTCOMES 243 // the total number of outcomes possible in wordle
#define GREEN_INPUT 'g'    // the green input character
#define ORANGE_INPUT 'o'   // the orange input character
#define BLACK_INPUT '-'    // the black input character

/**
 * Converts a character into it's lowercase equivalent.
 * @param character The character to convert.
 * @return The converted character.
 */
char to_lower(char character)
{
    if (character >= 'A' && character <= 'Z')
    {
        return character + 'a' - 'A';
    }
    return character;
}

/**
 * Get the first guess from the user's input.
 * @param guesses The valid guesses the user can choose from.
 * @return The user's first guess.
 */
char *get_first_guess(char guesses[][WORD_SIZE])
{
    static char first_guess[WORD_SIZE];
    printf("\n");
    while (1)
    {
        printf(" GUESS: ");

        for (int i = 0; i < WORD_SIZE - 1; i++) // get five characters
        {
            char c = to_lower(getchar());
            if (c == '\n') // skip end line character
            {
                i--;
                continue;
            }
            first_guess[i] = c;
        }
        first_guess[WORD_SIZE - 1] = '\0';

        for (int i = 0; i < GUESSES_LENGTH; i++)
        {
            if (strcmp(guesses[i], first_guess) == 0) // if we find the input in guesses return it
                return first_guess;
        }

        printf("** invalid first guess **\n"); // if we haven't found the input in guesses, print error message and retry
    }
}

/**
 * Get the result from the last guess from the user's input.
 * @return Pointer to the result array.
 */
int *get_result()
{
    static int result[WORD_SIZE - 1];
    while (1)
    {
        printf("RESULT: ");
        int error_detected = 0; // to detect when an invalid character is found
        for (int i = 0; i < WORD_SIZE - 1; i++)
        {
            char c = to_lower(getchar());
            if (c == '\n') // skip end line character
            {
                i--;
                continue;
            }

            if (c == GREEN_INPUT)
                result[i] = GREEN;
            else if (c == ORANGE_INPUT)
                result[i] = ORANGE;
            else if (c == BLACK_INPUT)
                result[i] = BLACK;
            else
                error_detected = 1; // invalid character was found
        }
        if (error_detected == 0)
            return result;                // if input was valid, return the result
        printf("** invalid result **\n"); // otherwise print error message and retry
    }
}

/**
 * Find the next best guess.
 * @param guesses The array of all valid guesses.
 * @param answers The array of all valid answers.
 * @param outcomes The array of possible outcomes.
 * @param answer_indexes The array of indexes of the available answers in the current game state.
 * @return The guess that minimises the average amount of answers left.
 */
char *find_best_guess(char guesses[GUESSES_LENGTH][WORD_SIZE], char answers[ANSWER_LENGTH][WORD_SIZE], int outcomes[TOTAL_OUTCOMES][WORD_SIZE - 1], int answer_indexes[ANSWER_LENGTH + 1])
{
    double min_average = ANSWER_LENGTH; // initialise minimum average
    static char min_word[WORD_SIZE];    // initialise minimum word

    // check all available answers first
    int z = 0;
    while (answer_indexes[z] != NULL_INDEX)
    {
        int total = TOTAL_OUTCOMES; // initialise total number of valid outcomes
        int possibles = 0;          // initialise number of possible answers

        for (int j = 0; j < TOTAL_OUTCOMES; j++) // for each outcome
        {
            int answers_left = 0; // initialise number of answers left for this outcome
            int k = 0;
            while (answer_indexes[k] != NULL_INDEX) // for each available answer check if it is possible with the current guess and outcome
            {
                if (is_possible(answers[answer_indexes[k]], answers[answer_indexes[z]], outcomes[j]))
                    answers_left++; // if so add to number of answers left for this outcome
                k++;
            }

            if (answers_left == 0)     // if we have no answers left for this outcome
                total--;               // valid outcomes decreases
            possibles += answers_left; // add to total number of possible answers
        }

        double average_words_left = (double)possibles / total; // get the average number of answers left for this guess
        if (average_words_left == 1)
            return answers[answer_indexes[z]]; // if average is 1, shortcut and return immediately (can't get better)

        if (average_words_left < min_average) // otherwise, update min_average, and min_word if out new average is smaller than min_average
        {
            strcpy(min_word, answers[answer_indexes[z]]);
            min_average = average_words_left;
        }

        if (z % 3 == 0)
            printf("Checking answers.\r"); // print current state of processing
        else if (z % 3 == 1)
            printf("Checking answers..\r");
        else
            printf("Checking answers...\r");
        fflush(stdout);
        z++;
    }

    for (int i = 0; i < GUESSES_LENGTH; i++) // for each guess
    {
        int total = TOTAL_OUTCOMES; // initialise total number of valid outcomes
        int possibles = 0;          // initialise number of possible answers

        for (int j = 0; j < TOTAL_OUTCOMES; j++) // for each outcome
        {
            int answers_left = 0; // initialise number of answers left for this outcome
            int k = 0;
            while (answer_indexes[k] != NULL_INDEX) // for each available answer check if it is possible with the current guess and outcome
            {
                if (is_possible(answers[answer_indexes[k]], guesses[i], outcomes[j]))
                    answers_left++; // if so add to number of answers left for this outcome
                k++;
            }

            if (answers_left == 0)     // if we have no answers left for this outcome
                total--;               // valid outcomes decreases
            possibles += answers_left; // add to total number of possible answers
        }

        double average_words_left = (double)possibles / total; // get the average number of answers left for this guess
        if (average_words_left == 1)
            return guesses[i]; // if average is 1, shortcut and return immediately (can't get better)

        if (average_words_left < min_average) // otherwise, update min_average, and min_word if out new average is smaller than min_average
        {
            strcpy(min_word, guesses[i]);
            min_average = average_words_left;
        }

        printf("Checking %d/%d\r", i + 1, GUESSES_LENGTH); // print current state of processing
        fflush(stdout);                                    // flush stdout for cleaner '\r' printing
    }

    return min_word; // return the best word found
}

/**
 * Filter the currently available answers to account for a new guess and result.
 * @param answers The array of all possible answers.
 * @param answer_indexes The indexes of the currently available answers for the current game state.
 * @param result The result of the new guess.
 * @param guess The new guess.
 * @return The total amount of viable answers left.
 */
int filter_answers(char answers[ANSWER_LENGTH][WORD_SIZE], int answer_indexes[ANSWER_LENGTH + 1], int result[WORD_SIZE - 1], char current_guess[WORD_SIZE])
{
    int new_indexes[ANSWER_LENGTH + 1]; // new indexes array
    int new_length = 0;
    int i = 0;

    while (answer_indexes[i] != NULL_INDEX) // loop through the currently available answers
    {
        int index = answer_indexes[i];
        if (is_possible(answers[index], current_guess, result)) // if the answer is still possible add to new indexes array
            new_indexes[new_length++] = index;
        i++;
    }
    new_indexes[new_length] = NULL_INDEX;

    int j = 0;
    while (new_indexes[j] != NULL_INDEX) // copy the new indexes array back into answer indexes
    {
        answer_indexes[j] = new_indexes[j];
        j++;
    }
    answer_indexes[j] = NULL_INDEX;

    return new_length; // return the total number of answers left
}

/**
 * Solve for a word.
 * @param total_guesses The total number of guesses so far.
 * @param current_guess The current guess for the word.
 * @param answers The array of all possible answers.
 * @param guesses The array of all possible guesses.
 * @param outcomes The array of all possible outcomes.
 * @param answer_indexes The array of indexes of all currently available answers in the game state.
 * @return The solution, or NULL if an error occurred.
 */
char *solve_word(int *total_guesses, char current_guess[WORD_SIZE], char answers[ANSWER_LENGTH][WORD_SIZE], char guesses[GUESSES_LENGTH][WORD_SIZE], int all_outcomes[TOTAL_OUTCOMES][WORD_SIZE - 1], int answer_indexes[ANSWER_LENGTH + 1])
{
    (*total_guesses)++;         // increment the total guesses
    int *result = get_result(); // get the result from the current guess

    int answers_left = filter_answers(answers, answer_indexes, result, current_guess); // filter the answers to find how many answers are left
    if (result[0] == GREEN && result[1] == GREEN && result[2] == GREEN && result[3] == GREEN && result[4] == GREEN)
    { // if the current guess is correct, return it and decrement total guesses (as we found it before the next guess)
        (*total_guesses)--;
        return current_guess;
    }
    else // otherwise print the amount of answers left
        printf("-- answers left: %d --\n\n", answers_left);

    if (answers_left == 0) // if we have no answers left, an error has occured so return null
        return NULL;
    else if (answers_left == 1) // if we have one answer left, return it
        return answers[answer_indexes[0]];

    char *next_best = find_best_guess(guesses, answers, all_outcomes, answer_indexes);           // otherwise find the next best guess
    printf(" GUESS: %s         \n", next_best);                                                  // print it out
    return solve_word(total_guesses, next_best, answers, guesses, all_outcomes, answer_indexes); // keep solving
}

int main()
{
    char answers[ANSWER_LENGTH][WORD_SIZE];
    char guesses[GUESSES_LENGTH][WORD_SIZE];
    get_answers_guesses(answers, guesses); // get answers and guesses

    char *first_guess = get_first_guess(guesses); // get the first guess

    int all_outcomes[TOTAL_OUTCOMES][WORD_SIZE - 1];
    get_all_possible_outcomes(all_outcomes); // get all outcomes

    int answer_indexes[ANSWER_LENGTH + 1]; // initialise answer indexes that stores the indexes of all currently available answers
    for (int i = 0; i < ANSWER_LENGTH; i++)
        answer_indexes[i] = i;
    answer_indexes[ANSWER_LENGTH] = NULL_INDEX;

    int total_guesses = 1;                                                                                    // we've had one guess already
    char *solution = solve_word(&total_guesses, first_guess, answers, guesses, all_outcomes, answer_indexes); // solve for the word

    if (solution == NULL) // if invalid results were inputted that lead to no answers left
    {
        printf("** ERROR: no answers left - invalid result(s) **\n");
        return 1;
    }
    printf("** ANSWER: %s - Guesses: %d **\n\n", solution, total_guesses); // print the summary of the solution
    return 0;
}
